### PR TITLE
added shared message library

### PR DESCRIPTION
**Pull request by smcgregor**
This pull request adds a shared messaging library meant to unify messaging between scripting contexts on different platforms.

Remaining todos:

* Get reviewed (@SummerWish)
* Debug selenium tests
* Add test coverage
* Add documentation
* Update documentation on priv.ly (@smcgregor)
* Integrate JetPack (Hitesh)
* Integrate Safari (Sambuddha)

### CODE REVIEW

File: shared/javascripts/context_messenger.js [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    // Don't send these messages to the extension context from the
+    // extension context.
+    if( Privly.message.currentContext === "BACKGROUND_SCRIPT" ) {
+      return;
+    }
+
+    // Platform specific messaging
+    var platform = Privly.message.currentPlatform;
+    if ( platform === "CHROME") {
+      chrome.extension.sendMessage(
```
Comment (SummerWish) [inline shared/javascripts/context_messenger.js]: `chrome.extension.sendMessage` is deprecated. we may use `chrome.runtime.sendMessage` instead : https://developer.chrome.com/extensions/extension

File: shared/javascripts/options.js [hunk 2]
```diff
@@ -160,20 +142,19 @@ if (Privly === undefined) {
...
-      if (request.ask.indexOf('options/') === 0) {
-        var method = request.ask.split('/')[1];
-        if (Privly.options[method] !== undefined) {
-          var returnValue = Privly.options[method].apply(Privly.options, request.params);
-          sendResponse(returnValue);
-        }
+  // Listen incoming messages to provide option interfaces
+  // for content scripts
+  function optionsChange(request) {
+    if (request.ask && request.ask.indexOf('options/') === 0) {
```
Comment (SummerWish) [inline shared/javascripts/options.js]: `request.action`

File: shared/javascripts/options.js [hunk 3]
```diff
@@ -69,30 +69,13 @@ if (Privly === undefined) {
...
-          });
-        });
-      }
-    }
+    var message = {
+      action: 'options/changed',
+      option: optionName,
+      newValue: optionValue
+    };
+    Privly.message.messageExtension(message);
```
Comment (SummerWish) [inline shared/javascripts/options.js]: we need to change `ask` in other place into `action`

File: shared/javascripts/options.js (hunk 2 above)
Comment (SummerWish) [inline shared/javascripts/options.js]: nevermind, it's ok

File: shared/javascripts/options.js [hunk 4]
```diff
@@ -160,20 +142,19 @@ if (Privly === undefined) {
...
+  }
 
-    // Listen incoming messages to provide option interfaces
-    // for content scripts
-    chrome.runtime.onMessage.addListener(function (request, sender, sendResponse) {
-      if (request.ask.indexOf('options/') === 0) {
-        var method = request.ask.split('/')[1];
-        if (Privly.options[method] !== undefined) {
-          var returnValue = Privly.options[method].apply(Privly.options, request.params);
-          sendResponse(returnValue);
```
Comment (SummerWish) [inline shared/javascripts/options.js]: missing `sendResponse`.
this interface allows both reading option values and write option values.

File: shared/javascripts/options.js (hunk 2 above)
Comment (SummerWish) [inline shared/javascripts/options.js]: we may need to unify those message payloads :-)

File: shared/javascripts/options.js [hunk 5]
```diff
@@ -160,20 +142,19 @@ if (Privly === undefined) {
...
-        }
+  // Listen incoming messages to provide option interfaces
+  // for content scripts
+  function optionsChange(request) {
```
Comment (SummerWish) [inline shared/javascripts/options.js]: `optionsChange` name may be confusing to others. this is an message-based function calling interface, specifically for content scripts.

File: shared/javascripts/options.js (hunk 3 above)
Comment (SummerWish) [inline shared/javascripts/options.js]: nevermind, all changed :-)

File: shared/javascripts/context_messenger.js [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+        message,
+        Privly.message.receiveMessage);
```
Comment (SummerWish) [inline shared/javascripts/context_messenger.js]: Is this code converts receiving response to receiving message? We may need to rewrite codes elsewhere (1), or provides response functionality (2) in the message library.. Currently many passively received response messages do not have a unified format or even an identifier. Message receiving handlers may not be able to handle those response messages.

here is my idea about (2):

Layer1: interface layer, we provide `sendMessage(data, responseCallback)` (or may we use Promise?) and `receiveMessage(callback<data, sendResponse(payload)>)`, similiar to the chrome message interface.
Layer2: response handler layer. this layer converts response messages to normal messages. it automatically add a unique identifier (and encapsule the payload) and send the response via normal message interface when `sendResponse` is called, and parse type=responses messages to forward them to the correct responseCallback.
Layer3: message compatible layer, provides unidirectional message sending and receiving functionality (already implemented)
we need a array to store those `responseCallback`. we may face memory leak issues if the developer specify a response callback but in fact he won't receive any responses.

Comment (SummerWish) [inline shared/javascripts/context_messenger.js]: pseudocode:
(original `Privly.message.messageExtension` renamed to `Privly.message._messageExtension`, `Privly.message.addListener` renamed to `Privly.message._addListener`)

```js
var identifierCounter = 0;

var resCallbacks = {};
var userListeners = [];

Privly.message.messageExtension = function(data, responseCallback) {
  var payload = {
    body: data,
    type: 'message',
    to: 'extension',
    from : '...',
    id: ++identifierCounter
  };
  Privly.message._messageExtension(data);
  if (typeof responseCallback === 'function') {
    resCallbacks[payload.id] = responseCallback;
  }
}

Privly.message._addListener(function (data) {
  if (data.type === 'message') {
    emitMessageListeners(data);
    return;
  }
  if (data.type === 'response') {
    if (resCallbacks[data.id] !== undefined) {
      resCallbacks[data.id](data.body);
      delete resCallbacks[data.id];
      return;
    }
  }
});

function emitMessageListeners(data) {
  userListeners.forEach(function (listener) {
    //.....
    listener(data.body, function (responseBody) {
      if (data.from === 'content_script') {
        Privly.message._messageContentScripts({
          body: responseBody,
          type: 'response',
          to: 'content_script',
          from: '...',
          id: data.id
        });
      } else if {
        //...
      }
    });
    //
  });
}

Privly.message.addListener = function(listener) {
  userListeners.push(listener);
};
```

Comment (SummerWish) [inline shared/javascripts/context_messenger.js]: a diagram for solution (2):

diagram

how do you think of this idea?

Comment (smcgregor) [inline shared/javascripts/context_messenger.js]: To my knowledge sendResponse is only available natively on Chrome extensions. Details for Jetpack are [here](https://developer.mozilla.org/en-US/Add-ons/SDK/Guides/Content_Scripts#Communicating_with_the_add-on). I don't think adding support for sendResponse should take too long, but I don't have time to implement it for about 2 weeks and two other students need something workable in the meantime. Is this something you want to solve before moving on with your posting work?

Comment (SummerWish) [inline shared/javascripts/context_messenger.js]: OK, I would try to send you a PR for this today

**Comment by SummerWish** [comment]
I am merging this PR now and I would make further changes.

**Comment by SummerWish** [comment]
@smcgregor send response functionality is implemented now. the rest issue is.. developer cannot sent message to a specific tab (and cannot know the message sender). embeded-posting application need this functionality to send message to its most outside context.

```
source -> target:

host page
--- the content script (target)
--- iframe1
--- --- iframe2
--- --- --- iframe_embed_posting
--- --- --- --- the app (source)
```

currently I use `source -> background script -> target` because the background script knows `tab.id` thus it can resend back the message.

I don't want to use `parent.poseMessage` because it would need to coordinate with several content scripts..

(seamless-posting doesn't need this feature)

**Comment by smcgregor** [comment]
How will you determine the target content script when you pass it through the background script?

**Comment by SummerWish** [comment]
The content script itself knowns whether it is the outmost (window.top) thus they can drop the message if it doesn't meet the requirement

**Pull request merged**